test/should_fail/poly_fail.erl

```diff
@@ -0,0 +1,25 @@
+-module(poly_fail).
+
+%% See "Bidirectional Typing for Erlang", N. Rajendrakumar, A. Bieniusa, 2021, Section 2. Examples.
+
+-export([find1/0,
+         poly_2/1,
+         poly_fail/3]).
+
+-gradualizer([solve_constraints]).
+
+-spec lookup(T1, [{T1, T2}]) -> (none | T2).
+lookup(_, []) -> none;
+lookup(K, [{K, V}|_]) -> V;
+lookup(K, [_|KVs])-> lookup(K, KVs).
+
+-spec find1() -> string().
+find1() ->
+    %% This should fail type checking.
+    "s" = lookup(0, [{0, 1}]).
+
+-spec poly_2(fun((A) -> A)) -> {integer(), boolean()}.
+poly_2(F) -> {F(42), F(false)}.
```

Comment:
This one shouldn't fail. Should it?

Reply:
I don't feel completely comfortable with this topic yet, but I think it depends on the future goals: if we ever want to be able to infer the type of a function like `poly_2` (for example, when it's used as an inline fun), then we're limited to rank-2 polymorphism for which inference is decidable.

According to _Bidirectional Typing for Erlang_:

> Erlang allows higher-ranked polymorphism where polymorphic types occur anywhere in the type signature. The Hindley-Milner type inference only supports the inference of rank-1 (prenex) polymorphism. Type inference for the higher-ranked polymorphism is undecidable, which is a significant limitation for the Hindley-Milner type inference. For example, the `poly_2` function shown next has rank-2 polymorphism that ETC fails to type-check:
>
>     -spec poly_2(fun((A) -> A)) -> {integer(), boolean()}.
>     poly_2(F) -> {F(42), F(false)}.
>
> Similarly, ETC fails for `poly_fail/3` due to its rank-2 poly-
morphism.
>
>     -spec poly_fail(fun((A) -> A), boolean(), integer()) -> {boolean(), integer()}.
>     poly_fail(F, B, I) -> {F(I), F(B)}.
>
> In contrast, Dialyzer here finds the return type mismatch error.

As we can see from this test, Gradualizer is also able to detect this error.

I'd leave it as is until anyone of us is comfortable enough with constraint solving and polymorphism to confidently move it to `should_pass` tests.